### DISCUSSION TITLE
Multi thread clients for connecting to different cluster

### DISCUSSION THREAD

**Discussion by hengsok** [josephg/node-foundationdb#60] [Q&A]
Hi @josephg, 

On Apple FDB documentation, it mentioned about multi threaded client for connecting to different clusters, could you help explain how can we set this option up or what is the code for connecting to multiple FDB clusters?  Thanks!
https://apple.github.io/foundationdb/api-general.html#multi-threaded-client

**Comment by josephg** [comment]
Hi!

It should work the same as other client libraries. Just connect to mulitple clusters. As for the multi-threaded client option, [try it out](https://github.com/josephg/node-foundationdb/blob/c1165539e5d4aea4765d8f6f6444a684156a1a21/lib/opts.g.ts#L202-L207). Though I'm not sure how much difference you'll see in performance, or what the implications are. You'll need to:

1. Set the FDB version
2. Use `fdb.configNetwork({client_threads_per_version: xxx})`
3. Open multiple clusters by making multiple db objects (using multiple calls to `open(clusterFileName)`

**Reply by hengsok** [reply]
Hi @josephg this is a very brilliant way to stay organised by having this discussion area. I was not sure where to ask questions too initially. Thanks so much for opening this space. Let me try your suggestions!

**Reply by hengsok** [reply]
In case someone else needs it, 

it turns out that if we turn on 'client_threads_per_version', we need to make use of external client library in order for it to work. If not we will get the error 'FDBError: Action not possible before the network is configured'. Below is the full working code. The location of external client library is pointing to the file on a mac. For linux, it would be '/usr/include/foundationdb/
libfdb_c.so'. I have also included in the config an example of how you can enable trace too which might be useful if you want to debug or find out about CPU usages.


```
const fdb = require('foundationdb');
fdb.setAPIVersion(630);
fdb.configNetwork({external_client_library: '/usr/local/lib/libfdb_c.dylib', client_threads_per_version: 2, trace_enable: '.', trace_format:'json'
});
```

**Reply by hengsok** [reply]
@josephg 

By the way, using this option has significant huge performance gain as I have analysed the trace file. It is in the order of 3X.

When we did not use multiple client threads while connecting to 2 clusters,
the 'MainThreadCPUSeconds' in trace shows 
MainThreadCPUSeconds => hover between 0.074-0.08

Whereas when we turn on 'client_threads_per_version:2',
MainThreadCPUSeconds => hover between 0.015-0.025

It can be concluded that the load is being load balanced between the 2 threads automatically by the fdb clients. It's really nice!

**Reply by josephg** [reply]
Awesome!

This is the first time I’ve used the discussion board here too. It’s great - I know what you mean about issues and Q&A stuff.

**Reply by hengsok** [reply]
Hope this thread is useful in getting people to get started with some of the config settings that are not documented yet regarding how to configure Network Config to use with external client. This can also be useful for configuring multi version clients.
